### PR TITLE
Properly initialize head time

### PR DESCRIPTION
**Pull request by fabxc**
This fixes various issues when initializing the head time range
under different starting conditions.

Depends on #332 

@krasi-georgiev as promised
@jkohen FYI as it depends on the new WAL stuff, but probably not too interesting for you

### CODE REVIEW

File: head.go [hunk 1]
```diff
@@ -462,10 +482,7 @@ func (h *Head) Truncate(mint int64) error {
 // for a compltely fresh head with an empty WAL.
```
Comment (krasi-georgiev) [inline head.go]: compltely -> completely

File: head.go [hunk 2]
```diff
@@ -576,15 +594,16 @@ func (h *Head) putBytesBuffer(b []byte) {
 }
 
 type headAppender struct {
-	head       *Head
-	mint, maxt int64
+	head         *Head
+	minValidTime int64 // No samples below this timestamp are allowed.
+	mint, maxt   int64
```
Comment (krasi-georgiev) [inline head.go]: Maybe some comments why we need separate `mint` and `minValidTime` ?

File: db_test.go [hunk 3]
```diff
@@ -1011,3 +1013,109 @@ func TestOverlappingBlocksDetectsAllOverlaps(t *testing.T) {
 		{Min: 8, Max: 9}: {nc1[8], nc1[9]},                                 // 7-10, 8-9
 	}, OverlappingBlocks(nc1))
 }
+
+func TestInitializeHeadTimestamp(t *testing.T) {
```
Comment (krasi-georgiev) [inline db_test.go]: shouldn't this test be in  `head_test.go`?

Comment (krasi-georgiev) [reply db_test.go]: actually there is already a head test where you can just add few lines similar to what I did in the other PR
https://github.com/prometheus/tsdb/pull/334/files#diff-0dc7b53c06c14aba584e5b4d6986fa07R96

Comment (fabxc) [reply db_test.go]: The test cover cases with existing blocks, which is beyond the scope of `Head`. `DB` has to startup correctly and make some calls of `Head`'s methods in the right order and with the right arguments. That's nothing we can test with an isolated `Head`.

File: head.go [hunk 4]
```diff
@@ -226,10 +226,27 @@ func (h *Head) processWALSamples(
...
+func (h *Head) updateMinMaxTime(mint, maxt int64) {
+	for {
+		lt := h.MinTime()
+		if mint >= lt {
+			break
+		}
+		if atomic.CompareAndSwapInt64(&h.minTime, lt, mint) {
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (fabxc) [reply head.go]: Because `CompareAndSwap` can fail and we have to retry it then.

[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply head.go]: thanks! I always thought these are blocking until they succeed.

File: db_test.go [hunk 5]
```diff
@@ -1011,3 +1013,109 @@ func TestOverlappingBlocksDetectsAllOverlaps(t *testing.T) {
...
+	t.Run("clean", func(t *testing.T) {
+		dir, err := ioutil.TempDir("", "test_head_init")
+		testutil.Ok(t, err)
+		defer os.RemoveAll(dir)
+
+		db, err := Open(dir, nil, nil, nil)
+		testutil.Ok(t, err)
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply db_test.go]: looking it again it is the same number of lines so all the same.

**Comment by fabxc** [comment]
Rebased and resolved conflicts.

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**